tests/Tokenizer/TokensTest.php

```diff
@@ -1856,6 +1856,32 @@ public function testFindingToken(): void
         self::assertTrue($tokens->isTokenKindFound(T_VARIABLE));
     }
 
+    public function testSetSizeCachePruning(): void
+    {
+        $tokens = Tokens::fromCode('<?php $a = function () {};');
+
+        self::assertSame(11, $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, 10));
+        self::assertSame(10, $tokens->findBlockStart(Tokens::BLOCK_TYPE_CURLY_BRACE, 11));
+        self::assertTrue($tokens->isTokenKindFound(T_FUNCTION));
+        self::assertTrue($tokens->isTokenKindFound('{'));
+        self::assertTrue($tokens->isTokenKindFound('}'));
+        self::assertTrue($tokens->isTokenKindFound(';'));
+        self::assertFalse($tokens->isTokenKindFound(T_CLASS));
+
+        $tokens->setSize(\count($tokens) - 2);
+
+        self::assertTrue($tokens->isTokenKindFound(T_FUNCTION));
+        self::assertTrue($tokens->isTokenKindFound('{'));
+        self::assertFalse($tokens->isTokenKindFound('}'));
```

Comment:
Are we sure this is correct? Changing the size doesn't change the fact that `}` was not removed from the collection.

I'd say the method `setSize` should not be called from outside of `Tokens` class. Or even, in our case not called at all and instead the custom implementation could be removed.

Reply:
I am 100% sure. `seSize()` method is public and once some elements are removed due optimized unset in `setSize()`, all internal caches must be adjusted (which this PR does newly) to be still valid.

Reply:
@kubawerlos can we please merge this PR?

Reply:
Rebased after #8257 - the `updateSize` method is newly internal, but it doesn't mean it must prune the cache correctly.